telemetry/telemetry.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package telemetry
+
+import (
+	"errors"
+	"fmt"
+	"go/ast"
+	"go/types"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+const (
+	telemetryPkgPath = "github.com/mattermost/mattermost-server/v5/services/telemetry"
+	modelPkgPath     = "github.com/mattermost/mattermost-server/v5/model"
+)
+
+// Analyzer describes telemetry status analysis function for a config setting
+var Analyzer = &analysis.Analyzer{
+	Name: "telemetry",
+	Doc:  "reports if a config field is not used in services/telemetry package",
+	Run:  run,
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	if pass.Pkg.Path() != telemetryPkgPath {
+		return nil, nil
+	}
+
+	// we need to find model.Config definition to parse its Type Spec
+	var configObject *types.TypeName
+	for _, obj := range pass.TypesInfo.Uses {
+		if tn, ok := obj.(*types.TypeName); ok && obj.Name() == "Config" && obj.Pkg().Path() == modelPkgPath {
+			configObject = tn
+			break
+		}
+	}
+
+	if configObject == nil {
+		return nil, errors.New("could not find model.Config type")
+	}
+
+	configMap, err := typeFieldMap(pass.Fset, configObject)
+	if err != nil {
+		return nil, fmt.Errorf("could generate fields map: %w", err)
+	}
+
+	var allExpressions []string
+
+	// We look every expression in the file so that we can
+	// check if the expected references exist in this package.
+
+	// NOTE: this approach does not check if the reference (config field) is added
```

Comment:
I *think* I'm comfortable with this approach, but 0/5 if others aren't.

A suggestion though, how about we move out the `trackConfig()` function into it's own file in the telemetry service? That way, the chance of the config field being used for something else in the file is much lower?

Reply:
It's doable, we can inspect a single file. Note that we also track plugins and we should also move things related to those as well.